pkg/component/runtime/manager_test.go

```diff
@@ -3481,6 +3481,164 @@ LOOP:
+				LogLevel: client.UnitLogLevelTrace,
+				Config: component.MustExpectedConfig(map[string]interface{}{
+					"type":    "fake",
+					"state":   int(client.UnitStateHealthy),
+					"message": "Fake Healthy 1",
+				}),
+			},
+			{
+				ID:       "fake-input-2",
+				Type:     client.UnitTypeInput,
+				LogLevel: client.UnitLogLevelTrace,
+				Config: component.MustExpectedConfig(map[string]interface{}{
+					"type":    "fake",
+					"state":   int(client.UnitStateHealthy),
+					"message": "Fake Healthy 2",
+				}),
+			},
+			{
+				ID:       "fake-input-3",
+				Type:     client.UnitTypeInput,
+				LogLevel: client.UnitLogLevelTrace,
+				Config: component.MustExpectedConfig(map[string]interface{}{
+					"type":    "fake",
+					"state":   int(client.UnitStateHealthy),
+					"message": "Fake Healthy 3",
+				}),
+			},
+			{
+				ID:       "fake-input-4",
+				Type:     client.UnitTypeInput,
+				LogLevel: client.UnitLogLevelTrace,
+				Config: component.MustExpectedConfig(map[string]interface{}{
+					"type":    "fake",
+					"state":   int(client.UnitStateHealthy),
+					"message": "Fake Healthy 4",
+				}),
+			},
+		},
+	}
+
+	subCtx, subCancel := context.WithCancel(context.Background())
+	defer subCancel()
+	subErrCh := make(chan error)
+	go func() {
+		sub := m.Subscribe(subCtx, "fake-default")
+		for {
+			select {
+			case <-subCtx.Done():
+				return
+			case state := <-sub.Ch():
+				t.Logf("component state changed: %+v", state)
+				if state.State == client.UnitStateFailed {
+					subErrCh <- fmt.Errorf("component failed: %s", state.Message)
+				} else {
+					healthyCount := 0
+					stoppedCount := 0
+					for _, unit := range state.Units {
+						if unit.State == client.UnitStateFailed {
+							subErrCh <- fmt.Errorf("unit failed: %s", unit.Message)
+						} else if unit.State == client.UnitStateHealthy {
+							healthyCount += 1
+						} else if unit.State == client.UnitStateStopped {
+							stoppedCount += 1
+						} else if unit.State == client.UnitStateStarting {
+							// acceptable
+						} else {
+							// unknown state that should not have occurred
+							subErrCh <- fmt.Errorf("unit reported unexpected state: %v", unit.State)
+						}
+					}
+					if healthyCount == 4 {
+						// remove the component which will stop it
+						m.Update(component.Model{Components: []component.Component{}})
+						err := <-m.errCh
+						if err != nil {
+							subErrCh <- err
+						}
+					} else if stoppedCount == 4 {
+						subErrCh <- nil
+					}
+				}
+			}
+		}
+	}()
+
+	defer drainErrChan(errCh)
+	defer drainErrChan(subErrCh)
+
+	m.Update(component.Model{Components: []component.Component{comp}})
+	err = <-m.errCh
+	require.NoError(t, err)
+
+	endTimer := time.NewTimer(30 * time.Second)
+	defer endTimer.Stop()
+LOOP:
+	for {
+		select {
+		case <-endTimer.C:
+			t.Fatalf("timed out after 30 seconds")
+		case err := <-errCh:
+			require.NoError(t, err)
+		case err := <-subErrCh:
+			require.NoError(t, err)
```

Comment:
[Suggestion]
Use `Errorf` and add a message to better differentiate both error cases

Reply:
Each error that comes over those channels are specific and its clear. Adding more context would be redundant.